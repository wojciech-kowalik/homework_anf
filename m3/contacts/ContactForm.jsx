/* eslint-disable */
import React from 'react';

export class ContactForm extends React.Component {

  static defaultProps = {
    contact: {
      name: '',
      details: ''
    }
  }

  onChange = (e) => {
    let target = e.target;
    let name = target.name;
    let value = target.type === "checkbox" ? target.checked : target.value;

    this.props.onChange({
      ...this.props.contact,
      [name]: value
    });
  }

  onSubmit = (e) => {
    e.preventDefault();
    this.props.onSubmit(this.props.contact);
  }

  render() {
    return (
      <form onSubmit={this.onSubmit}>
        <div className="form-group">
          <label className="d-block w-100">
            Name:
            <input
              className="form-control"
              name="name"
              value={this.props.contact.name}
              onChange={this.onChange}
            />
          </label>
        </div>
        <div className="form-group">
          <label className="d-block w-100">
            Contact details:
            <textarea
              className="form-control"
              name="details"
              value={this.props.contact.details}
              onChange={this.onChange}
            />
          </label>
        </div>
        <div className="form-group">
          <input
            type="button"
            id="cancel-contact"
            className="btn btn-default"
            value="Cancel"
            onClick={this.props.onCancel}
          />
          <input
            type="submit"
            id="save-contact"
            className="btn btn-primary ml-2"
            value="Save"
          />
        </div>
      </form>
    );
  }
}
