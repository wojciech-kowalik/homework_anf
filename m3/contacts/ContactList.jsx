/* eslint-disable */
import React from 'react';

export class ContactList extends React.Component{

  static defaultProps = {
    contacts: []
  }

  render() {
    return (
      <div className="list-group">
        {this.props.contacts.map((contact) => (
          <button
            key={contact.id}
            onClick={(e) => this.props.onSelect(contact)}
            className={`list-group-item ${this.props.selected && (this.props.selected.id === contact.id) ? 'active' : ''}`}
          >
            {contact.name}
          </button>
        ))}
      </div>
    );
  }
}
