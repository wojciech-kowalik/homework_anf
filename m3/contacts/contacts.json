[
	{
		"id": "ad324356-384a-4fa4-a1d9-e373b0f760d0",
		"name": "Allison Clayton",
		"details": "383-897-142"
	},
	{
		"id": "cbe348af-a5ae-47c2-8d2b-07441a773fb7",
		"name": "David Jacobs",
		"details": "david.jacobs@erntogra.com"
	},
	{
		"id": "4822e046-eee3-4d3a-b9be-f860f48ea5df",
		"name": "Dominic Ramsey",
		"details": "117 Aster Court APT 187\nMiami 78499"
	},
	{
		"id": "3d3a4d97-0bad-4ea7-93e0-e936dcc02928",
		"name": "Victoria Clarke",
		"details": "victoria.clarke@buapel.com"
	},
	{
		"id": "4bd7bd06-c5c5-428f-bde9-de07e59628e7",
		"name": "Zoe M. Russell",
		"details": "zoe.russell@furba.biz\n014-203-3635"
	}
]
