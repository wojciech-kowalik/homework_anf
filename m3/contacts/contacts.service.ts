import { v4 as uuid } from "uuid";

export interface Contact {
  id: string;
  name: string;
  details: string;
}

export interface ServiceInterface<T> {
  getAll: () => Promise<T[]>;
  update: (updateCallback: (data: T[]) => void) => void;
  save: (entity: T) => Promise<T>;
}

export class ContactsService implements ServiceInterface<Contact> {
  constructor(private contacts: Contact[]) {}

  async getAll() {
    return this.contacts;
  }

  update(updateCallback: (contacts: Contact[]) => void) {
    const clone = [...this.contacts];
    updateCallback(clone);
    this.contacts = clone;
  }

  async save(contact: Contact) {
    const index = this.contacts.findIndex(({ id }) => contact.id === id);

    if (index === -1) {
      contact.id = uuid();
      this.update((contacts) => contacts.push(contact));
    } else {
      this.update((contacts) => contacts.splice(index, 1, contact));
    }

    return contact;
  }
}
