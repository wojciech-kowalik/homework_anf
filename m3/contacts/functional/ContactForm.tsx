import React from "react";
import { Contact } from "../contacts.service";

interface ContactFormProps {
  contact: Contact;
  onChange: (contact: Contact) => void;
  onSubmit: (contact: Contact) => void;
  onCancel: () => void;
}

export const ContactForm: React.FC<ContactFormProps> = (props): JSX.Element => {
  const { contact, onChange, onSubmit, onCancel } = props;

  const _onChange = (
    e: React.ChangeEvent<HTMLInputElement & HTMLTextAreaElement>
  ) => {
    const target = e.target;
    const name = target.name;
    const value = target.type === "checkbox" ? target.checked : target.value;

    onChange({
      ...contact,
      [name]: value,
    });
  };

  const _onSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    onSubmit(contact);
  };

  return (
    <form onSubmit={_onSubmit}>
      <div className="form-group">
        <label className="d-block w-100">
          Name:
          <input
            className="form-control"
            name="name"
            data-testid="form-name"
            value={contact.name}
            onChange={_onChange}
          />
        </label>
      </div>
      <div className="form-group">
        <label className="d-block w-100">
          Contact details:
          <textarea
            className="form-control"
            name="details"
            data-testid="form-details"
            value={contact.details}
            onChange={_onChange}
          />
        </label>
      </div>
      <div className="form-group">
        <input
          type="button"
          id="cancel-contact"
          data-testid="form-button-cancel"
          className="btn btn-default"
          value="Cancel"
          onClick={onCancel}
        />
        <input
          type="submit"
          id="save-contact"
          data-testid="form-button-submit"
          className="btn btn-primary ml-2"
          value="Save"
        />
      </div>
    </form>
  );
};
