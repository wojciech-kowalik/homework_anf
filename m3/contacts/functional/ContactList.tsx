import React from "react";
import { Contact } from "../contacts.service";

interface ContactListProps {
  contacts: Contact[];
  selected: Contact | null;
  onSelect: (contact: Contact) => void;
}

export const ContactList: React.FC<ContactListProps> = (props): JSX.Element | null => {
  const { contacts, selected, onSelect } = props;

  if (contacts.length === 0) {
    return null;
  }

  return (
    <div className="list-group">
      {contacts.map((contact) => (
        <button
          key={contact.id}
          data-testid={`button-${contact.id}`}
          onClick={(e) => onSelect(contact)}
          className={`list-group-item ${
            selected && selected.id === contact.id ? "active" : ""
          }`}
        >
          {contact.name}
        </button>
      ))}
    </div>
  );
};
