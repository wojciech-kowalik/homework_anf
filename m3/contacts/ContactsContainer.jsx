/* eslint-disable */
import React from 'react';

import { ContactList } from './ContactList';
import { ContactForm } from './ContactForm';

export class ContactsContainer extends React.Component {

  constructor(props) {
    super(props);

    // ContactsService object
    this.service = this.props.service;

    this.state = {
      contacts: [],
      selected: null
    };

    this.service.getAll().then((contacts) => {
      this.setState({
        contacts
      });
    });
  }

  newContact = () => {
    this.setState({
      selected: {
        name: '',
        details: ''
      }
    });
  }

  onSelect = (contact) => {
    this.setState({
      selected: contact
    });
  }

  onSubmit = (contact) => {
    this.service.save(contact).then(()=>{
      this.service.getAll().then((contacts) => {
        this.setState({
          contacts
        });
      });
    });
  }

  onCancel = () => {
    this.setState({
      selected: null
    });
  }

  render() {
    return (
      <div className="container">
        <div className="row">
          <div className="col-md-12">
            <h1>Contacts</h1>
          </div>
        </div>
        <div className="row">
          <div className="col-md-4">
            <ContactList
              contacts={this.state.contacts}
              selected={this.state.selected}
              onSelect={this.onSelect}
            />
          </div>
          <div className="col-md-4">
            {this.state.selected
            ? (
              <ContactForm contact={this.state.selected}
                onChange={this.onSelect}
                onSubmit={this.onSubmit}
                onCancel={this.onCancel}
              />
            ) : (
              <div>
                <button id="new-contact" onClick={this.newContact} className="btn btn-primary">New contact</button>
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}
