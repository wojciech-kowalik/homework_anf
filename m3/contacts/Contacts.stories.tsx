/* eslint-disable import/first */
import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Meta } from '@storybook/react/types-6-0';

import contacts from './contacts.json';
import { ContactsContainer as FunctionalContainer } from './functional/ContactsContainer';
import { ContactsContainer } from "./ContactsContainer";
import { ContactsService } from './contacts.service';

const svc = new ContactsService(contacts);

import { lessons } from 'stories';
export default {
  title: lessons.m3.add('HOMEWORK').toString(),
  argTypes: {
  },
} as Meta;

export const Contacts = () => {
  return (
    <>
      <link
        rel="stylesheet"
        href="https://maxcdn.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css"
        integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO"
        crossOrigin="anonymous"
      />
      <ContactsContainer service={svc} />
      <FunctionalContainer service={svc} />
    </>
  );
}
