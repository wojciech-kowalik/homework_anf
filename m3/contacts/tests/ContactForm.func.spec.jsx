import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import contacts from "../contacts.json";
import { ContactForm } from "../functional/ContactForm";

describe("ContactForm Component", () => {
  it("should be stateless component (no setState, only props)", () => {
    render(<ContactForm contact={{}} />);
    expect(screen.getByTestId("form-name").value).toBe("");
    expect(screen.getByTestId("form-details").value).toBe("");
  });

  it("should match snapshot", () => {
    const { container } = render(<ContactForm contact={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should display name and details form input fields after a contact is selected", () => {
    const contact = contacts[1];
    render(<ContactForm contact={contact} />);
    expect(screen.getByTestId("form-name").value).toEqual(contact.name);
    expect(screen.getByTestId("form-details").value).toEqual(contact.details);
  });

  it("should call onChange with changed form values", async () => {
    const onChange = jest.fn();
    const contact = { name: "", details: "" };
    render(<ContactForm contact={contact} onChange={onChange} />);

    fireEvent.change(screen.getByTestId("form-name"), {
      target: { value: "test" },
    });
    //await userEvent.type(input, 'test'); // with userEvent calls 4 times
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith({ name: "test", details: "" });

    fireEvent.change(screen.getByTestId("form-details"), {
      target: { value: "test" },
    });
    expect(onChange).toBeCalledTimes(2);
    expect(onChange).toBeCalledWith({ name: "", details: "test" });
  });

  it("should call onSubmit with changed contact after the form is submitted", async () => {
    const onSubmit = jest.fn();
    const contact = { name: "", details: "" };
    render(<ContactForm contact={contact} onSubmit={onSubmit} />);

    await userEvent.click(screen.getByTestId("form-button-submit"));
    expect(onSubmit).toBeCalledTimes(1);
    expect(onSubmit).toBeCalledWith(contact);
  });
});
