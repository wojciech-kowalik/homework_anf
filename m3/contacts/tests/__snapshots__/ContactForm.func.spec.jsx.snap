// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm Component should match snapshot 1`] = `
<div>
  <form>
    <div
      class="form-group"
    >
      <label
        class="d-block w-100"
      >
        Name:
        <input
          class="form-control"
          data-testid="form-name"
          name="name"
          value=""
        />
      </label>
    </div>
    <div
      class="form-group"
    >
      <label
        class="d-block w-100"
      >
        Contact details:
        <textarea
          class="form-control"
          data-testid="form-details"
          name="details"
        />
      </label>
    </div>
    <div
      class="form-group"
    >
      <input
        class="btn btn-default"
        data-testid="form-button-cancel"
        id="cancel-contact"
        type="button"
        value="Cancel"
      />
      <input
        class="btn btn-primary ml-2"
        data-testid="form-button-submit"
        id="save-contact"
        type="submit"
        value="Save"
      />
    </div>
  </form>
</div>
`;
